providers/azure.go

```diff
@@ -177,21 +177,27 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
 	return session, nil
 }
 
-// EnrichSession finds the email to enrich the session state
+// EnrichSession finds the email and groups claim to enrich the session state
 func (p *AzureProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
-	if s.Email != "" {
-		return nil
+	if s.Email == "" {
+		// add email
+		email, err := p.getEmailFromProfileAPI(ctx, s.AccessToken)
+		if err != nil {
+			return fmt.Errorf("unable to get email address: %v", err)
+		}
+		if email == "" {
+			return errors.New("unable to get email address")
+		}
+		s.Email = email
 	}
 
-	email, err := p.getEmailFromProfileAPI(ctx, s.AccessToken)
-	if err != nil {
-		return fmt.Errorf("unable to get email address: %v", err)
-	}
-	if email == "" {
-		return errors.New("unable to get email address")
+	if s.Groups == nil {
+		// add groups claim
+		oidcClaims, err := p.verifyTokenAndExtractClaims(ctx, s.IDToken)
```

Comment:
You are verifying & extracting the same token/claims twice it appears (once for email another for groups)?

I think this should be refactored to just handle email & groups (if applicable) from the same extracted claims. OIDCProvider should have samples of this pattern I imagine.

This is also not properly handling any non-nil errors (which likely won't happen because they would happen in the call above for emails 😄 ) -- so unifying will clear up the logic and gaps I think.

Reply:
Scratch that - I mistook `getEmailFromProfileURL` for `verifyTokenAndExtractEmail` -- but that is also called in the chain, so I think the double call concern stands.

Reply:
I think you want to emulate the `createSession` helper method in the OIDCProvider that will build the user/email/groups from the claims and is used wherever needed (initial redeem, refreshes).

`EnrichSession` is more intended for network calls to fetch any missing data only in the initial login.

If you moved all the claims logic together in the `createSession` parallel, you'd also get the missing functionality of setting the group from claims in refresh calls (which I believe this is missing at the moment?)

Reply:
The current `verifyTokenAndExtractEmail` is attached in the proper places, I think you want the `group` claim extraction to happen there.

Reply:
I think if we were to reorder this a bit, we could avoid calling it twice. For example, check at first if both email and groups are empty, return early, if not, we need the oidcClaims to populate at least one of them, then perform the rest of the logic with only fetching it once

I had intended to make a generic utility for extracting claims like this, a `claimExtractor` kind of thing that would be able to used to get a named claim, and would only do the verification if there was a need to actually get a token, or could also look at the profile URL if needed. Maybe this would be a good time to do that.